generate simpler expressions for sin, cos, tan, cot(k*pi/n) , etc, for s...

postvakje opened the pull request:
...ome values of n

For n=12, 15, 40, and some other values, generate expressions with
smaller number of nested sqrt.   For example currently cos(pi/12),
tan(pi/8), and cot(pi/12) return 
sqrt(sqrt(3)/4 + 1/2), 
sqrt(-sqrt(2)/4 + 1/2)/sqrt(sqrt(2)/4 + 1/2) and 
sqrt(sqrt(3)/4 + 1/2)/sqrt(-sqrt(3)/4 + 1/2) resp.  
The current patch returns 
sqrt(2)/4 + sqrt(6)/4,
sqrt(2)*(-sqrt(2)/2 + 1) and 
sqrt(3) + 2 resp.

smichr commented on sympy/functions/elementary/tests/test_trigonometric.py:
```diff
@@ -542,7 +558,7 @@ def test_cot_rewrite():
     assert cot(tan(x)).rewrite(
         exp).subs(x, 3).n() == cot(x).rewrite(exp).subs(x, tan(3)).n()
     assert cot(log(x)).rewrite(Pow) == -I*(x**-I + x**I)/(x**-I - x**I)
-    assert cot(4*pi/15).rewrite(pow) == (cos(4*pi/15)/sin(4*pi/15)).rewrite(pow)
```
I don't think you should delete the old test, should you?

smichr commented on sympy/functions/elementary/trigonometric.py:
```diff
@@ -814,6 +839,31 @@ def eval(cls, arg):
+                    if not isinstance(cresult, cos) \
+                            and not isinstance(sresult, cos):
+                        if sresult == 0:
+                            return S.ComplexInfinity
```
Could you add a test that covers this line? While you are editing, could you also add space around the `-` in the next line (and in other places that I mark with "pep8")?

smichr commented on sympy/functions/elementary/trigonometric.py:
```diff
@@ -502,8 +507,26 @@ def eval(cls, arg):
+                    nvala, nvalb = cls(a), cls(b)
+                    if None == nvala or None == nvalb:
+                        return None
+                    return nvala*nvalb+cls(S.Pi/2-a)*cls(S.Pi/2-b)
```
pep8

smichr commented on sympy/functions/elementary/trigonometric.py:
```diff
@@ -814,6 +839,31 @@ def eval(cls, arg):
+                    nvala, nvalb = cls(p*S.Pi/table2[q][0]), cls(p*S.Pi/table2[q][1])
+                    if None == nvala or None == nvalb:
+                        return None
+                    return (nvala-nvalb)/(1+nvala*nvalb)
```
pep8

smichr commented on sympy/functions/elementary/trigonometric.py:
```diff
@@ -1047,6 +1099,31 @@ def eval(cls, arg):
+                            and not isinstance(sresult, cos):
+                        if sresult == 0:
+                            return S.ComplexInfinity
+                        return (1+cresult)/sresult
```
pep8

smichr commented on sympy/functions/elementary/trigonometric.py:
```diff
@@ -1047,6 +1099,31 @@ def eval(cls, arg):
+                    nvala, nvalb = cls(p*S.Pi/table2[q][0]), cls(p*S.Pi/table2[q][1])
+                    if None == nvala or None == nvalb:
+                        return None
+                    return (1+nvala*nvalb)/(nvalb-nvala)
```
pep8

smichr commented:
None of the tests you added involve a return value of zoo. (And I'm not sure that all those additional tests are necessary since if you select 1 value from the dictionary you should be able to select any of them.) Is it possible to cover that line that I mentioned?

postvakje commented:
The only time that line gets triggered when evaluating tan(pi/n) is when sin(2*pi/n) = 0, i.e. when n = 2, 2/3, etc.  And the cases of n = 2 and n = 2/3 has been covered by the tests 
    assert tan(pi/2) == zoo
    assert tan(3*pi/2) == zoo
in test_trigonometric.py

smichr commented:
Hmmm...not sure why the coverage report showed that as not being covered. But I see with a print statement that you are right. I'll check again. In the meanwhile, could you squash this all together into a single commit?

smichr commented on sympy/functions/elementary/trigonometric.py:
```diff
@@ -1047,6 +1099,31 @@ def eval(cls, arg):
+                    if not isinstance(cresult, cos) \
+                            and not isinstance(sresult, cos):
+                        if sresult == 0:
+                            return S.ComplexInfinity
```
ahh...it's this one that's not covered.

postvakje replied:
I guess that line never gets called, since if q > 2 then 2/q < 1 and sin(2*pi/q) != 0.  So I can get rid of the lines:
```python
if sresult == 0:
    return S.ComplexInfinity
``` 

I will also squash the commits

smichr commented:
If you run into difficulties generating the single commit, please see [this note](https://github.com/sympy/sympy/pull/8721#issuecomment-71766695) that I wrote in another PR.

postvakje commented:
I hope I did it correctly.  I created the squashed commit (19677f3) and merge it with the branch on the server into 0d094f5.

smichr commented:
> I hope I did it correctly

There are still 10 commits (last I looked). I think if you follow the guide I gave that will resolve things for you. If you are nervous about ruining anything, just make a back up before you start (`git branch backup-name`). Just ask if you run into difficulties.

postvakje committed:
trigonometric: generate simpler expressions for trigfunctions(k*pi/n)

For n = 12, 15, 40, and some other values, generate expressions of
sin, cos, tan, cot(k*pi/n) with smaller number of nested sqrt.
For example currently cos(pi/12), tan(pi/8), and cot(pi/12) return
sqrt(sqrt(3)/4 + 1/2) , sqrt(-sqrt(2)/4 + 1/2)/sqrt(sqrt(2)/4 + 1/2)
and sqrt(sqrt(3)/4 + 1/2)/sqrt(-sqrt(3)/4 + 1/2) resp.
The current patch returns sqrt(2)/4 + sqrt(6)/4, sqrt(2)*(-sqrt(2)/2 + 1)
and sqrt(3) + 2 resp.

* simplify tan(k*pi/2n) and cot(k*pi/2n) for values of n such that
  explicit expression exists.
* added additional tests to test_trigonometric and changed tests in numbers,
  geometry and pretty
* Since tan(pi/15) and cot(pi/15) are evaluated by current patch, the tests
for them are moved from test_tan_rewrite to test_tan and from
test_cot_rewrite to test_cot.  Also the tests are changed somewhat since
even though cos(pi/15)*tan(pi/15)  and sin(pi/15) return mathematically
equivalent equations, they do not cancel out under simplify.

smichr commented:
OK :-) Now please use `git commit --amend` to add a [commit message](https://github.com/sympy/sympy/wiki/Development-workflow#writing-commit-messages).

smichr commented:
Your title is too long...please make it conform to [the guide](https://github.com/sympy/sympy/wiki/Development-workflow) -- see the section "Writing commit messages". I tried to give a direct link to that section [in my comment to you](https://github.com/sympy/sympy/pull/8878#issuecomment-72368887) but for some reason it jumps to the wrong section of the document. If you select the table of contents section yourself it should work.  (And if you or @asmeurer knows why that is broken or what I did wrong, please let me know.)

smichr commented:
+1

smichr commented:
Thanks, this is in.
